### ISSUE TITLE
Create app-target for each executable cabal package

### ISSUE THREAD

**Issue by HariAmoor-professional** [srid/haskell-flake#36]
This ticket corresponds to @roberth's [comment](https://github.com/srid/haskell-flake/pull/34#discussion_r1010415656) on #34. To restate, we need to create a flake output in `apps'.*` for each executable cabal package. The logic to filter executable packages from non-executable ones is non-trivial.

**Comment by srid** [comment]
@roberth @HariAmoor-professional Why do we need this? `nix run` already looks in `packages.*` if apps are not defined.

**Comment by roberth** [comment]
I think it's relevant for multi-executable packages and packages whose executable has a different name than the package name.

**Comment by srid** [comment]
How do we approach this?

- Proper way: parse the `.cabal` file, and extract executable stanzas.
- Quick-and-dirty way: look under `${drv}/bin` and use those executables.

**Comment by roberth** [comment]
> parse the `.cabal` file

A regex on the cabal file contents may actually go a long way.
Isn't something like `^executable ([a-zA-Z0-9_.-]*)$` on any line a decent way to find all executables? Haven't checked the grammar, but this probably catches over 99% in practice.


> * look under `${drv}/bin`

We shouldn't do IFD to evaluate `attrNames apps`. That would slow down all apps evaluations, even if they're not from a haskell project.

**Comment by srid** [comment]
Yea, we should parse the `.cabal` file. 

What should the interface be? In the context of #62, to begin with I think we could add an option under `outputs`, like `config.haskellProjects.default.outputs.localApps` that points to the apps themselves. Or do we want to expose them as per-package sets, like `outputs.localPackageInfo.<package-name>.<exe-name> = <nix-app>`?

**Comment by roberth** [comment]
> EDIT: Or we can put this in the `passthru` of packages?


Putting it in `passthru` merges it with an IFD-based attrset, incurring the IFD cost, so that's not good enough for generating an apps attrset.

That shouldn't stop us from also adding information from passthru, although perhaps consistency with the rest of the ecosystem is a reason to make such improvements in cabal2nix instead.


> outputs.localPackageInfo.<package-name>.<exe-name> = <nix-app>

Here everything except `<nix-app>` is computable without IFD or even evaluating any packages. If some code evaluates up to `<exe-name>` ignoring the values, it will be instant. This format does not lose any information learned from the cabal files, and is unopinionated. I think it's a good idea to have this information around in an option. Something like `localApps` seems a bit more "opinionated" as it needs to handle name collisions or decide what the app names should be. Nonetheless, it's likely that the default behavior will be good enough, so this is also worth adding.
tl;dr both?

**Comment by srid** [comment]
> Here everything except `<nix-app>` is computable without IFD or even evaluating any packages. If some code evaluates up to `<exe-name>` ignoring the values, it will be instant. This format does not lose any information learned from the cabal files, and is unopinionated. I think it's a good idea to have this information around in an option.

So,  let's go with the `outputs.localPackageInfo.<package-name>.<exe-name>` approach while documenting the caveats of evaluating the values. I suppose in the future, we could store more than the executable metadata here; [haskell.nix](https://input-output-hk.github.io/haskell.nix/dev/dev-architecture.html#packages) exports a bunch of things from the cabal file:

<img width="576" alt="image" src="https://user-images.githubusercontent.com/3998/229520635-051dac68-6e28-4584-b2c5-ae89c50c7376.png">

So maybe (for future expansion): `outputs.localPackageInfo.<package-name>.exes.<exe-name>` 

---

Do we still need a separate `outputs.localPackages` and `outputs.localPackageInfo`? Why not merge them both, à ls:

```nix
outputs.localPackages = {
  foo = {
    package = <the drv>;
    exes.foo-exe = <the app>;
  };
}
```

**Comment by roberth** [comment]
Good points, and doing it all merged in `localPackages` sgtm.

I confused `packages` and `localPackages` for a sec. If you add it to `packages`, the implementation of the reading part becomes local to the package submodule. Now I wonder if we have a reason why `localPackages.<name>` isn't `packages.<name>.package`. That's a bit off-topic though.

**Comment by srid** [comment]
> Now I wonder if we have a reason why `localPackages.<name>` isn't `packages.<name>.package`. That's a bit off-topic though.

We can do it as part of this change (#137). It would also be consistent with the `haskellProjects.<name>.packages` option naming. So we have, `basePackages` + `packages` = `finalPackages`. Not sure if those are the best names, though. Or maybe we should just call it `localPackages` through out.
